runtime/src/accounts_db.rs

```diff
@@ -749,6 +810,40 @@ impl AccountsDB {
         infos
     }
 
+    pub fn verify_hash_internal_state(&self, fork: Fork, ancestors: &HashMap<Fork, usize>) -> bool {
+        let mut hash_state = BankHash::default();
+        let hashes: Vec<_> = self.scan_accounts(
+            ancestors,
+            |collector: &mut Vec<BankHash>, option: Option<(&Pubkey, Account, Fork)>| {
+                if let Some((pubkey, account, fork)) = option {
+                    if !sysvar::check_id(&account.owner) {
+                        let hash = BankHash::from_hash(&Self::hash_account(fork, &account, pubkey));
+                        debug!("xoring..{} key: {}", hash, pubkey);
+                        collector.push(hash);
+                    }
+                }
+            },
+        );
+        for hash in hashes {
+            Self::xor_hash_state(&mut hash_state, hash);
+        }
+        let fork_hashes = self.fork_hashes.read().unwrap();
+        if let Some((_, state)) = fork_hashes.get(&fork) {
+            hash_state == *state
+        } else {
+            false
+        }
+    }
+
+    pub fn xor_in_hash_state(&self, fork_id: Fork, hash: BankHash) {
+        let mut fork_hashes = self.fork_hashes.write().unwrap();
+        let fork_hash_state = fork_hashes
+            .entry(fork_id)
+            .or_insert((false, BankHash::default()));
```

Comment:
or_insert() should only ever happen on slot 0, right?

I'm reading that bank hash chaining is also an XOR? not a hash of a hash?

should we assert that fork_hash_state.0 is false somewhere in here, or possibly skip re-doing the XOR if we've already done so?

Reply:
> or_insert() should only ever happen on slot 0, right?

That's right.

> I'm reading that bank hash chaining is also an XOR? not a hash of a hash?

Yes. It's a hash of the resulting image which is from the xor of all the live accounts states.

> should we assert that fork_hash_state.0 is false somewhere in here, or possibly skip re-doing the XOR if we've already done so?

Well this would be called multiple times per slot on every store, so the flag will be set the first time and then be true after that. The flag is just for keeping track of when no accounts have been xor'ed yet so that we can return `None` for the `hash_internal_state` api. It seemed cleaner to start that as the parent hash value so then there is no special casing when we xor in the new accounts states.

Reply:
ok, thanks for explaining